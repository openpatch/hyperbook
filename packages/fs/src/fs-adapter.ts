/**
 * Filesystem abstraction layer for cross-platform compatibility
 * Supports both Node.js and VS Code Web environments
 */

export interface FileSystemAdapter {
  // Async operations
  readFile(path: string, encoding?: BufferEncoding): Promise<string>;
  readFile(path: string, encoding: null): Promise<Buffer>;
  writeFile(path: string, data: string | Buffer): Promise<void>;
  readdir(path: string): Promise<string[]>;
  stat(path: string): Promise<FileStats>;
  exists(path: string): Promise<boolean>;

  // Sync operations (may not be available in all environments)
  readFileSync(path: string, encoding?: BufferEncoding): string;
  readFileSync(path: string, encoding: null): Buffer;
  writeFileSync(path: string, data: string | Buffer): void;
  existsSync(path: string): boolean;
  rmSync?(path: string): void;  // Optional, for cleanup operations
}

export interface FileStats {
  isDirectory(): boolean;
  isFile(): boolean;
}

export interface PathAdapter {
  join(...paths: string[]): string;
  resolve(...paths: string[]): string;
  relative(from: string, to: string): string;
  dirname(path: string): string;
  basename(path: string, ext?: string): string;
  extname(path: string): string;
  parse(path: string): { dir: string; name: string; ext: string; base: string; root: string };
  sep: string;
  posix: {
    join(...paths: string[]): string;
  };
}

let currentFsAdapter: FileSystemAdapter | null = null;
let currentPathAdapter: PathAdapter | null = null;

export function setFileSystemAdapter(adapter: FileSystemAdapter) {
  currentFsAdapter = adapter;
}

export function setPathAdapter(adapter: PathAdapter) {
  currentPathAdapter = adapter;
}

export function getFileSystemAdapter(): FileSystemAdapter {
  if (!currentFsAdapter) {
    throw new Error('FileSystem adapter not initialized. Call setFileSystemAdapter() first.');
  }
  return currentFsAdapter;
}

export function getPathAdapter(): PathAdapter {
  if (!currentPathAdapter) {
    throw new Error('Path adapter not initialized. Call setPathAdapter() first.');
  }
  return currentPathAdapter;
}
