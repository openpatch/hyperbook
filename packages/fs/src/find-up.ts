/**
 * Custom find-up implementation for Hyperbook
 * Searches for a file or directory by walking up parent directories
 * Uses the filesystem adapter for cross-platform compatibility
 */

import { getFileSystemAdapter, getPathAdapter } from "./fs-adapter";

export interface FindUpOptions {
  cwd?: string;
  type?: "file" | "directory";
}

/**
 * Asynchronously find a file or directory by walking up parent directories
 * @param name - The name of the file or directory to find
 * @param options - Options for the search
 * @returns The absolute path to the found file/directory, or undefined if not found
 */
export async function findUp(
  name: string,
  options: FindUpOptions = {}
): Promise<string | undefined> {
  const fs = getFileSystemAdapter();
  const path = getPathAdapter();
  
  const { cwd = process.cwd(), type } = options;
  let currentDir = path.resolve(cwd);
  
  // Keep track of visited directories to avoid infinite loops
  const visited = new Set<string>();
  
  while (true) {
    // Avoid infinite loops
    if (visited.has(currentDir)) {
      return undefined;
    }
    visited.add(currentDir);
    
    const candidatePath = path.join(currentDir, name);
    
    try {
      const stat = await fs.stat(candidatePath);
      
      // Check if the type matches (if specified)
      if (type === "file" && stat.isFile()) {
        return candidatePath;
      } else if (type === "directory" && stat.isDirectory()) {
        return candidatePath;
      } else if (!type) {
        // If no type specified, return any match
        return candidatePath;
      }
    } catch (error) {
      // File/directory doesn't exist at this level, continue searching
    }
    
    // Move to parent directory
    const parentDir = path.dirname(currentDir);
    
    // If we've reached the root (parent is same as current), stop
    if (parentDir === currentDir) {
      return undefined;
    }
    
    currentDir = parentDir;
  }
}

/**
 * Synchronously find a file or directory by walking up parent directories
 * @param name - The name of the file or directory to find
 * @param options - Options for the search
 * @returns The absolute path to the found file/directory, or undefined if not found
 */
export function findUpSync(
  name: string,
  options: FindUpOptions = {}
): string | undefined {
  const fs = getFileSystemAdapter();
  const path = getPathAdapter();
  
  const { cwd = process.cwd(), type } = options;
  let currentDir = path.resolve(cwd);
  
  // Keep track of visited directories to avoid infinite loops
  const visited = new Set<string>();
  
  while (true) {
    // Avoid infinite loops
    if (visited.has(currentDir)) {
      return undefined;
    }
    visited.add(currentDir);
    
    const candidatePath = path.join(currentDir, name);
    
    try {
      // Check if file/directory exists
      if (!fs.existsSync(candidatePath)) {
        // Move to parent directory
        const parentDir = path.dirname(currentDir);
        
        // If we've reached the root (parent is same as current), stop
        if (parentDir === currentDir) {
          return undefined;
        }
        
        currentDir = parentDir;
        continue;
      }
      
      // For sync operations, we need to determine if it's a file or directory
      // We can do this by trying to read it as a directory
      if (type === "file") {
        // For file type, we just check existence (already done above)
        return candidatePath;
      } else if (type === "directory") {
        // For directory type, try to read it as a directory
        try {
          fs.readFileSync(candidatePath, "utf8");
          // If readFile succeeds, it's a file, not a directory
          // Move to parent
          const parentDir = path.dirname(currentDir);
          if (parentDir === currentDir) {
            return undefined;
          }
          currentDir = parentDir;
          continue;
        } catch (e) {
          // If readFile fails, it might be a directory
          // This is a heuristic - in sync mode, we have limited options
          return candidatePath;
        }
      } else {
        // If no type specified, return any match
        return candidatePath;
      }
    } catch (error) {
      // File/directory doesn't exist at this level, continue searching
      const parentDir = path.dirname(currentDir);
      
      // If we've reached the root (parent is same as current), stop
      if (parentDir === currentDir) {
        return undefined;
      }
      
      currentDir = parentDir;
    }
  }
}
