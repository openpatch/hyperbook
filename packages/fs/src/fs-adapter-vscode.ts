/**
 * VS Code Web filesystem adapter
 * Uses vscode.workspace.fs API for browser-compatible file operations
 */
import { FileSystemAdapter, PathAdapter, FileStats } from "./fs-adapter";

// These will be provided by the VS Code extension context
let vscodeWorkspaceFs: any = null;
let vscodeUri: any = null;

export function setVSCodeWorkspaceFs(fs: any, Uri?: any) {
  vscodeWorkspaceFs = fs;
  if (Uri) {
    vscodeUri = Uri;
  }
}

class VSCodeFileStats implements FileStats {
  constructor(private type: number) {}

  isDirectory(): boolean {
    // FileType.Directory = 2
    return this.type === 2;
  }

  isFile(): boolean {
    // FileType.File = 1
    return this.type === 1;
  }
}

function uriFromPath(path: string): any {
  if (!vscodeWorkspaceFs) {
    throw new Error('VS Code workspace.fs not initialized');
  }
  // Use vscode.Uri if available, otherwise create a simple object
  if (vscodeUri && vscodeUri.file) {
    return vscodeUri.file(path);
  }
  // Fallback for simple path handling
  return { fsPath: path, path };
}

export const vscodeFileSystemAdapter: FileSystemAdapter = {
  async readFile(path: string, encoding?: BufferEncoding | null): Promise<any> {
    if (!vscodeWorkspaceFs) {
      throw new Error('VS Code workspace.fs not initialized');
    }
    
    const uri = uriFromPath(path);
    const buffer = await vscodeWorkspaceFs.readFile(uri);
    
    if (encoding === null) {
      return buffer;
    }
    
    const decoder = new TextDecoder(encoding || 'utf-8');
    return decoder.decode(buffer);
  },

  async writeFile(path: string, data: string | Buffer): Promise<void> {
    if (!vscodeWorkspaceFs) {
      throw new Error('VS Code workspace.fs not initialized');
    }
    
    const uri = uriFromPath(path);
    const buffer = typeof data === 'string' 
      ? new TextEncoder().encode(data)
      : new Uint8Array(data);
    
    return vscodeWorkspaceFs.writeFile(uri, buffer);
  },

  async readdir(path: string): Promise<string[]> {
    if (!vscodeWorkspaceFs) {
      throw new Error('VS Code workspace.fs not initialized');
    }
    
    const uri = uriFromPath(path);
    const entries = await vscodeWorkspaceFs.readDirectory(uri);
    return entries.map(([name]: [string, number]) => name);
  },

  async stat(path: string): Promise<FileStats> {
    if (!vscodeWorkspaceFs) {
      throw new Error('VS Code workspace.fs not initialized');
    }
    
    const uri = uriFromPath(path);
    const stat = await vscodeWorkspaceFs.stat(uri);
    return new VSCodeFileStats(stat.type);
  },

  async exists(path: string): Promise<boolean> {
    try {
      await this.stat(path);
      return true;
    } catch {
      return false;
    }
  },

  readFileSync(path: string, encoding?: BufferEncoding | null): any {
    // Sync operations are not available in VS Code Web
    throw new Error('Synchronous file operations are not supported in VS Code Web environment');
  },

  writeFileSync(path: string, data: string | Buffer): void {
    // Sync operations are not available in VS Code Web
    throw new Error('Synchronous file operations are not supported in VS Code Web environment');
  },

  existsSync(path: string): boolean {
    // Sync operations are not available in VS Code Web
    throw new Error('Synchronous file operations are not supported in VS Code Web environment');
  },

  rmSync(path: string): void {
    // Sync operations are not available in VS Code Web
    throw new Error('Synchronous file operations are not supported in VS Code Web environment');
  },
};

// Path utilities for web (browser-compatible)
export const vscodePathAdapter: PathAdapter = {
  join(...paths: string[]): string {
    const normalized = paths
      .filter(p => p && p.length > 0)
      .join('/')
      .replace(/\/+/g, '/');
    return normalized;
  },

  resolve(...paths: string[]): string {
    let resolvedPath = '';
    
    for (let i = paths.length - 1; i >= 0; i--) {
      const path = paths[i];
      if (!path || path.length === 0) continue;
      
      if (path.startsWith('/')) {
        resolvedPath = path;
        break;
      }
      
      resolvedPath = path + (resolvedPath ? '/' + resolvedPath : '');
    }
    
    return resolvedPath || '/';
  },

  relative(from: string, to: string): string {
    from = from.replace(/\\/g, '/');
    to = to.replace(/\\/g, '/');
    
    const fromParts = from.split('/').filter(p => p);
    const toParts = to.split('/').filter(p => p);
    
    let i = 0;
    while (i < fromParts.length && i < toParts.length && fromParts[i] === toParts[i]) {
      i++;
    }
    
    const upCount = fromParts.length - i;
    const relativeParts = Array(upCount).fill('..');
    const remainingParts = toParts.slice(i);
    
    return [...relativeParts, ...remainingParts].join('/');
  },

  dirname(path: string): string {
    path = path.replace(/\\/g, '/');
    const lastSlash = path.lastIndexOf('/');
    if (lastSlash === -1) return '.';
    if (lastSlash === 0) return '/';
    return path.substring(0, lastSlash);
  },

  basename(path: string, ext?: string): string {
    path = path.replace(/\\/g, '/');
    const lastSlash = path.lastIndexOf('/');
    let base = lastSlash === -1 ? path : path.substring(lastSlash + 1);
    
    if (ext && base.endsWith(ext)) {
      base = base.substring(0, base.length - ext.length);
    }
    
    return base;
  },

  extname(path: string): string {
    path = path.replace(/\\/g, '/');
    const lastDot = path.lastIndexOf('.');
    const lastSlash = path.lastIndexOf('/');
    
    if (lastDot === -1 || lastDot < lastSlash) return '';
    
    return path.substring(lastDot);
  },

  parse(path: string): { dir: string; name: string; ext: string; base: string; root: string } {
    path = path.replace(/\\/g, '/');
    const lastSlash = path.lastIndexOf('/');
    const lastDot = path.lastIndexOf('.');
    
    const dir = lastSlash === -1 ? '' : path.substring(0, lastSlash);
    const base = lastSlash === -1 ? path : path.substring(lastSlash + 1);
    const ext = (lastDot !== -1 && lastDot > lastSlash) ? path.substring(lastDot) : '';
    const name = ext ? base.substring(0, base.length - ext.length) : base;
    const root = path.startsWith('/') ? '/' : '';
    
    return { dir, name, ext, base, root };
  },

  sep: '/',

  posix: {
    join(...paths: string[]): string {
      return paths
        .filter(p => p && p.length > 0)
        .join('/')
        .replace(/\/+/g, '/');
    },
  },
};
