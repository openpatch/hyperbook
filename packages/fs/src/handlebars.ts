import handlebars from "handlebars";
import { lookup } from "mime-types";
import { getFileSystemAdapter, getPathAdapter } from "./fs-adapter";
import { extractLines, VFileBase } from "./vfile";
import { findUpSync } from "./find-up";

const fs = () => getFileSystemAdapter();
const path = () => getPathAdapter();

declare global {
  var workspaceRoot: string;
}

const isString = (s: any): s is string => {
  return typeof s === "string";
};

const chop = (str: string): string => {
  if (!isString(str)) return "";
  var re = /^[-_.\W\s]+|[-_.\W\s]+$/g;
  return str.trim().replace(re, "");
};

const changecase = (str: string, fn: (s: string) => string): string => {
  if (!isString(str)) return "";
  if (str.length === 1) {
    return str.toLowerCase();
  }

  str = chop(str).toLowerCase();

  var re = /[-_.\W\s]+(\w|$)/g;
  return str.replace(re, function (_, ch) {
    return fn(ch);
  });
};

const registerHelpers = (handlebars: any, options?: { file: VFileBase }) => {
  const cwd = options?.file.root ?? process.cwd();
  handlebars.registerHelper("times", (n: number, block: any) => {
    let accum = "";
    for (let i = 0; i < n; ++i) accum += block.fn(i);
    return accum;
  });

  handlebars.registerHelper("concat", (...s: string[]) => {
    return s.filter(s => typeof s !== "object").join("");
  });

  handlebars.registerHelper("camelcase", (s: string) => {
    if (!isString(s)) return "";
    return changecase(s, (c) => c.toUpperCase());
  });

  handlebars.registerHelper("dashcase", (s: string) => {
    if (!isString(s)) return "";
    return changecase(s, (c) => "-" + c);
  });

  handlebars.registerHelper("lowercase", (s: string) => {
    if (!isString(s)) return "";
    return s.toLowerCase();
  });

  handlebars.registerHelper("uppercase", (s: string) => {
    if (!isString(s)) return "";
    return s.toUpperCase();
  });

  handlebars.registerHelper("pascalcase", (s: string) => {
    if (!isString(s)) return "";
    s = changecase(s, (c) => c.toUpperCase());
    return s.charAt(0).toUpperCase() + s.slice(1);
  });

  handlebars.registerHelper("replaceAll", (s: string, a: string, b: string) => {
    if (!isString(s)) return "";
    if (!isString(a)) return s;
    if (!isString(b)) b = "";
    return s.replaceAll(a, b);
  });

  handlebars.registerHelper("replace", (s: string, a: string, b: string) => {
    if (!isString(s)) return "";
    if (!isString(a)) return s;
    if (!isString(b)) b = "";
    return s.replace(a, b);
  });

  handlebars.registerHelper("replace", (s: string, a: string, b: string) => {
    if (!isString(s)) return "";
    if (!isString(a)) return s;
    if (!isString(b)) b = "";
    return s.replace(a, b);
  });

  handlebars.registerHelper("rbase64", (src: string) => {
    let gitRoot = findUpSync(".git", { type: "file", cwd: cwd });
    if (!gitRoot) {
      gitRoot = findUpSync(".git", { type: "directory", cwd: cwd });
      if (!gitRoot) {
        return `rbase64 is only applicable in git projects. No .git was found in ${cwd} and above.`;
      }
    }
    let p = path().join(path().dirname(gitRoot), src);
    try {
      const fileDataBase64 = fs().readFileSync(p, "base64");
      const mime = lookup(p);
      return `data:${mime};base64,${fileDataBase64}`;
    } catch (e) {
      return `File at ${src} is missing.`;
    }
  });

  handlebars.registerHelper(
    "rfile",
    (src: string, lines?: string, ellipsis?: string) => {
      if (!src) {
        throw Error("file needs a path to a file");
      }
      let gitRoot = findUpSync(".git", { type: "file", cwd: cwd });
      if (!gitRoot) {
        gitRoot = findUpSync(".git", {
          type: "directory",
          cwd: cwd,
        });
        if (!gitRoot) {
          return `rfile is only applicable in git projects. No .git was found in ${cwd} and above`;
        }
      }
      let p = path().join(path().dirname(gitRoot), src);
      try {
        const content = fs().readFileSync(p, "utf8");
        return extractLines(content, lines, ellipsis);
      } catch (e) {
        return `File ${src} is missing.`;
      }
    },
  );

  handlebars.registerHelper("base64", (src: string) => {
    let p = path().join(cwd, src);
    const fileDataBase64 = fs().readFileSync(p, "base64");
    const mime = lookup(p);
    return `data:${mime};base64,${fileDataBase64}`;
  });

  handlebars.registerHelper(
    "file",
    (src: string, lines?: string, ellipsis?: string) => {
      if (!src) {
        throw Error("file needs a path to a file");
      }
      let p = path().join(cwd, src);
      const content = fs().readFileSync(p, "utf8");
      return extractLines(content, lines, ellipsis);
    },
  );
};

export { registerHelpers, handlebars };
