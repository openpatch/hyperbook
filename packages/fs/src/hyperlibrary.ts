import { getFileSystemAdapter, getPathAdapter } from "./fs-adapter";
const fs = () => getFileSystemAdapter();
const path = () => getPathAdapter();

import { HyperlibraryJson } from "@hyperbook/types";
import { findUp } from "find-up";

export const getJson = async (root: string): Promise<HyperlibraryJson> => {
  return fs()
    .readFile(path().join(root, "hyperlibrary.json"))
    .then((f: string) => f)
    .then(JSON.parse);
};

export const find = async (file: string): Promise<HyperlibraryJson> => {
  return findUp("hyperlibrary.json", {
    cwd: file,
  } as any)
    .then((f) => {
      if (!f) {
        throw new Error("Could not find hyperlibrary.json");
      }
      return fs().readFile(f);
    })
    .then((f) => JSON.parse(f.toString()));
};

export const findRoot = async (file: string): Promise<string> => {
  return findUp("hyperlibrary.json", {
    cwd: file,
  } as any).then((f) => {
    if (!f) {
      throw new Error("Could not find hyperlibrary.json");
    }
    return path().parse(f).dir;
  });
};
