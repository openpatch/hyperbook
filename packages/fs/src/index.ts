import fs from "fs/promises";
import path from "path";
import matter from "gray-matter";
import chalk from "chalk";
import {
  HyperbookFrontmatter,
  HyperbookJson,
  Hyperproject,
  HyperlibraryJson,
  Language,
  Link,
  Navigation,
  HyperbookPage,
  HyperbookSection,
  Glossary,
} from "@hyperbook/types";
import { findUp } from "find-up";

const listFiles = async (
  root: string,
  dir: string = "",
  absolute: boolean = true
): Promise<string[]> => {
  const files = await fs.readdir(path.join(root, dir));
  return Promise.all(
    files.flatMap(async (file) => {
      const stat = await fs.stat(path.join(root, dir, file));
      if (stat.isDirectory()) {
        return listFiles(root, path.join(dir, file), absolute);
      } else if (file.endsWith(".md")) {
        if (absolute) {
          return [path.join(root, dir, file)];
        } else {
          return [path.join(dir, file)];
        }
      }
      return [];
    })
  ).then((f) => f.flat());
};

export const readBook = async (root: string, absolute: boolean = true) => {
  return listFiles(path.join(root, "book"), "", absolute);
};

export const readGlossary = async (root: string, absolute: boolean = true) => {
  return listFiles(path.join(root, "glossary"), "", absolute);
};

export const makeGlossary = async (root: string): Promise<Glossary> => {
  const files = await readGlossary(root);
  const glossary: Glossary = {};

  for (const file of files) {
    const { content, data } = await readFile(file);
    let name = path.basename(file, ".md");
    if (data.name) {
      name = data.name;
    } else {
      console.log(
        `\n${chalk.yellow(
          `warn  `
        )}- Glossary page ${file} does not specify a name. Defaulting to the filename ${name}.`
      );
    }

    const letter = name[0].toUpperCase();
    if (!glossary[letter]) {
      glossary[letter] = [];
    }

    const relativePath = path
      .relative(root, file)
      .replace(/\.md$/, "")
      .split("/");

    glossary[letter].push({
      name,
      href: "/" + relativePath.join("/"),
    });
  }

  return glossary;
};

export const listPagesForTerm = async (
  root: string,
  term: string
): Promise<HyperbookPage[]> => {
  const files = await readBook(root);

  const pages: HyperbookPage[] = [];

  for (const file of files) {
    const { content, data } = await readFile(file);
    const r = new RegExp(`:t\\[.*\\]\\{#${term}(\..*)?\\}|:t\\[${term}\\]`);
    const m = content.match(r);

    if (m && !data.hide && data.name) {
      const relativePath = path
        .relative(root, file)
        .replace(/\.md$/, "")
        .split("/");
      const isIndex = relativePath[relativePath.length - 1] === "index";
      if (isIndex) {
        relativePath.pop();
      }
      pages.push({
        ...data,
        href: "/" + relativePath.join("/"),
      });
    }
  }

  return pages;
};

export const readFile = async (file: string) => {
  const source = await fs
    .readFile(file)
    .catch(() =>
      fs
        .readFile(file + ".md")
        .catch(() => fs.readFile(path.join(file, "index.md")))
    );

  const { content, data } = matter(source);

  return { content, data: data as HyperbookFrontmatter };
};

export const readHyperbook = async (root: string) => {
  return fs
    .readFile(path.join(root, "hyperbook.json"))
    .then((f) => f.toString())
    .then(JSON.parse) as Promise<HyperbookJson>;
};

export const readHyperlibrary = async (root: string) => {
  return fs
    .readFile(path.join(root, "hyperlibrary.json"))
    .then((f) => f.toString())
    .then(JSON.parse) as Promise<HyperlibraryJson>;
};

export const findHyperbook = async (file: string): Promise<HyperbookJson> => {
  return findUp("hyperbook.json", {
    cwd: file,
  } as any)
    .then((f) => {
      if (!f) {
        throw new Error("Could not find hyperbook.json");
      }
      return fs.readFile(f);
    })
    .then((f) => JSON.parse(f.toString()));
};

export const findHyperbookRoot = async (file: string): Promise<string> => {
  return findUp("hyperbook.json", {
    cwd: file,
  } as any).then((f) => {
    if (!f) {
      throw new Error("Could not find hyperbook.json");
    }
    return path.parse(f).dir;
  });
};

export const readProject = async (
  root: string,
  libraryEntry?: HyperlibraryJson["library"][0]
): Promise<Hyperproject> => {
  if (libraryEntry?.src) {
    root = path.join(root, libraryEntry.src);
  }
  const hyperbookJson = await readHyperbook(root).catch(() => null);
  if (hyperbookJson) {
    return {
      type: "book",
      src: root,
      basePath: libraryEntry?.basePath ?? hyperbookJson.basePath,
      name: libraryEntry?.name ?? hyperbookJson.name,
      icon: libraryEntry?.icon,
    };
  }

  const hyperlibraryJson = await readHyperlibrary(root).catch(() => null);
  if (hyperlibraryJson) {
    return {
      type: "library",
      src: root,
      basePath: libraryEntry?.basePath ?? hyperlibraryJson.basePath,
      name: libraryEntry?.name ?? hyperlibraryJson.name,
      projects: await Promise.all(
        hyperlibraryJson.library.map((p) =>
          readProject(root, {
            ...p,
            basePath: path.join(
              libraryEntry?.basePath ?? hyperlibraryJson.basePath ?? "",
              p.basePath
            ),
          })
        )
      ),
      icon: libraryEntry?.icon,
    };
  }

  console.log(
    `${chalk.red("Error")} - Missing book or library for path ${root}.`
  );

  throw Error(`Missing book or library for path ${root}`);
};

export const getProjectName = (project: Hyperproject, language?: Language) => {
  let label = "";
  if (typeof project.name === "string") {
    label = project.name;
  } else {
    if (language) {
      label = project.name[language] || "";
    } else {
      label = Object.values(project.name)[0];
    }
    if (!label) {
      console.log(
        chalk.red(
          `You need to provide a name for language ${language} in ${project.src}`
        )
      );
      throw Error("");
    }
  }
  return label;
};

type MakeLinkForHyperprojectOptions = {
  href?: {
    useSrc?: true;
    append?: string[];
    prepend?: string[];
    relative?: string;
    protocol?: string;
  };
};

export const makeLinkForHyperproject = async (
  project: Hyperproject,
  language: Language = "en",
  options: MakeLinkForHyperprojectOptions = {}
): Promise<Link> => {
  const label = getProjectName(project, language);

  if (project.type === "library") {
    return {
      label,
      links: await Promise.all(
        project.projects.map((p) =>
          makeLinkForHyperproject(p, language, options)
        )
      ),
      icon: project.icon,
    };
  } else {
    let href = project.basePath ?? "/";
    if (!href.startsWith("/")) {
      href = "/" + href;
    }

    if (options.href?.useSrc) {
      href = project.src;
    }
    if (options.href?.relative) {
      href = path.relative(options.href.relative, href);
    }
    if (options.href?.prepend) {
      href = path.join(...options.href.prepend, href);
    }
    if (options.href?.append) {
      href = path.join(href, ...options.href.append);
    }
    if (options.href?.protocol) {
      if (href.startsWith("/")) {
        href = href.slice(0, -1);
      }
      href = options.href.protocol + href;
    }

    if (href.length > 1 && href.endsWith("/")) {
      href = href.slice(0, -1);
    }

    return {
      label,
      href,
      icon: project.icon,
    };
  }
};

export const makeNavigationForHyperbook = async (
  root: string,
  currPath: string = "/"
): Promise<Navigation> => {
  const hyperbook = await readHyperbook(root);

  const getSectionsAndPages = async function (
    dirPath: string,
    pageList: HyperbookPage[] = []
  ) {
    const files = await fs.readdir(dirPath);
    let arrayOfPages: HyperbookPage[] = [];
    let arrayOfSections: HyperbookSection[] = [];

    for (const file of files) {
      let p = path.join(dirPath, file);
      let repo: string | null = null;
      if (hyperbook.repo) {
        repo = hyperbook.repo + "/" + path.relative(root, p);
      }

      const stat = await fs.stat(p);

      if (stat.isDirectory()) {
        const { pages, sections } = await getSectionsAndPages(p, pageList);
        const { content, data } = await readFile(p);
        const section: HyperbookSection = {
          ...data,
          href: "/" + path.relative(path.join(root, "book"), p),
          isEmpty: content.trim() === "",
          pages,
          sections,
        };
        if (repo) {
          section.repo = repo + "/index.md";
        }

        arrayOfSections.push(section);
      } else {
        const { data } = await readFile(p);
        if (p.endsWith(".md")) {
          p = p.substring(0, p.length - 3);
          if (path.relative(path.join(root, "book"), p) === "index") {
            p = p.substring(0, p.length - 5);
          }
          if (!p.endsWith("index")) {
            const page: HyperbookPage = {
              ...data,
              href: "/" + path.relative(path.join(root, "book"), p),
            };
            if (repo) {
              page.repo = repo;
            }

            arrayOfPages.push(page);
          }
        }
      }
    }

    arrayOfPages = arrayOfPages.sort((a, b) => (a.name > b.name ? 1 : -1));
    arrayOfPages = arrayOfPages.sort((a, b) => {
      const iIndex = a.index !== undefined ? a.index : 9999;
      const eIndex = b.index !== undefined ? b.index : 9999;
      return iIndex - eIndex;
    });
    arrayOfSections = arrayOfSections.sort((a, b) =>
      a.name > b.name ? 1 : -1
    );
    arrayOfSections = arrayOfSections.sort((a, b) => {
      const iIndex = a.index !== undefined ? a.index : 9999;
      const eIndex = b.index !== undefined ? b.index : 9999;
      return iIndex - eIndex;
    });

    return { pages: arrayOfPages, sections: arrayOfSections };
  };

  const getPageList = (
    sections: HyperbookSection[],
    pages: HyperbookPage[]
  ): HyperbookPage[] => {
    let pageList = [...pages];

    for (const section of sections) {
      pageList = [
        ...pageList,
        section,
        ...getPageList(section.sections, section.pages),
      ];
    }

    return pageList;
  };

  const { sections, pages } = await getSectionsAndPages(
    path.join(root, "book")
  );

  let pageList = getPageList(sections, pages);

  let i = pageList.findIndex((p) => p.href === currPath);
  const current = pageList[i] || null;

  pageList = pageList.filter((p) => !p.isEmpty && !p.hide);
  i = pageList.findIndex((p) => p.href === currPath);

  const next = pageList[i + 1] || null;
  const previous = pageList[i - 1] || null;

  return {
    next,
    current,
    previous,
    sections,
    pages,
  };
};
