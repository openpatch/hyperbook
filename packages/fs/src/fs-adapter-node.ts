/**
 * Node.js filesystem adapter
 */
import fs from "fs/promises";
import fsSync from "fs";
import nodePath from "path";
import { FileSystemAdapter, PathAdapter, FileStats } from "./fs-adapter";

class NodeFileStats implements FileStats {
  constructor(private stats: fsSync.Stats) {}

  isDirectory(): boolean {
    return this.stats.isDirectory();
  }

  isFile(): boolean {
    return this.stats.isFile();
  }
}

export const nodeFileSystemAdapter: FileSystemAdapter = {
  async readFile(path: string, encoding?: BufferEncoding | null): Promise<any> {
    if (encoding === null) {
      return fs.readFile(path);
    }
    return fs.readFile(path, encoding || "utf8");
  },

  async writeFile(path: string, data: string | Buffer): Promise<void> {
    return fs.writeFile(path, data as any);
  },

  async readdir(path: string): Promise<string[]> {
    return fs.readdir(path);
  },

  async stat(path: string): Promise<FileStats> {
    const stats = await fs.stat(path);
    return new NodeFileStats(stats);
  },

  async exists(path: string): Promise<boolean> {
    try {
      await fs.access(path);
      return true;
    } catch {
      return false;
    }
  },

  readFileSync(path: string, encoding?: BufferEncoding | null): any {
    if (encoding === null) {
      return fsSync.readFileSync(path);
    }
    return fsSync.readFileSync(path, encoding || "utf8");
  },

  writeFileSync(path: string, data: string | Buffer): void {
    fsSync.writeFileSync(path, data as any);
  },

  existsSync(path: string): boolean {
    return fsSync.existsSync(path);
  },

  rmSync(path: string): void {
    fsSync.rmSync(path);
  },
};

export const nodePathAdapter: PathAdapter = {
  join: nodePath.join.bind(nodePath),
  resolve: nodePath.resolve.bind(nodePath),
  relative: nodePath.relative.bind(nodePath),
  dirname: nodePath.dirname.bind(nodePath),
  basename: nodePath.basename.bind(nodePath),
  extname: nodePath.extname.bind(nodePath),
  parse: nodePath.parse.bind(nodePath),
  sep: nodePath.sep,
  posix: {
    join: nodePath.posix.join.bind(nodePath.posix),
  },
};
