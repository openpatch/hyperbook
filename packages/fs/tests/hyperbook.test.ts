import path from "path";
import { describe, it, expect } from "vitest";
import { hyperbook, vfile } from "../src";

describe("hyperbook", () => {
  const relative = (s: string) =>
    path.relative(path.join(__dirname, "fixtures"), s);
  const makeFileRelative = (f: vfile.VFile) => {
    if ("references" in f) {
      return {
        ...f,
        root: relative(f.root),
        path: { ...f.path, absolute: relative(f.path.absolute) },
        elements: [],
        references: f.references.map((f) => ({
          ...f,
          root: relative(f.root),
          path: { ...f.path, absolute: relative(f.path.absolute) },
        })),
      } as vfile.VFileGlossary;
    }
    return {
      ...f,
      root: relative(f.root),
      path: { ...f.path, absolute: relative(f.path.absolute) },
    };
  };
  it("should get navigation", async () => {
    let hyperbookPath = path.join(__dirname, "fixtures", "single-hyperbook");
    let files = await vfile.list(hyperbookPath);
    files = files.map(makeFileRelative);
    let current = files.find(
      (f) => f.name === "paradigms" && f.folder === "book",
    );
    if (!current) {
      throw Error("Missing file");
    }
    const pagesAndSections = await hyperbook.getPagesAndSections(hyperbookPath);
    const pageList = hyperbook.getPageList(
      pagesAndSections.sections,
      pagesAndSections.pages,
    );
    const navigation = await hyperbook.getNavigationForFile(pageList, current);
    expect(navigation).toMatchSnapshot();
  });
});
