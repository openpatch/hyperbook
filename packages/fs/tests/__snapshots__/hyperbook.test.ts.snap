// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hyperbook > should get navigation 1`] = `
{
  "current": {
    "href": "/paradigms",
    "name": "Paradigms",
    "repo": "https://github.com/openpatch/hyperbook/edit/main/website/en/book/paradigms.md",
  },
  "next": {
    "href": "/hyperbook-test",
    "name": "hyperbook-test",
    "repo": "https://github.com/openpatch/hyperbook/edit/main/website/en/book/hyperbook-test.md",
  },
  "previous": {
    "href": "/",
    "name": "Home",
    "repo": "https://github.com/openpatch/hyperbook/edit/main/website/en/book/index.md",
  },
}
`;
